### PR TITLE
bugfix #2550

### PR DESCRIPTION
**Pull request by junpenglao**
- [x] Add support for`n` as a vector input in random method
- [x] Add test

### CODE REVIEW

**Comment by twiecki** [comment]
LGTM.

**Comment by junpenglao** [comment]
Did some local testing after https://github.com/pymc-devs/pymc3/pull/2540, seems using `self.shape` is not very reliable - will fix that here instead.

File: pymc3/distributions/multivariate.py [hunk 1]
```diff
@@ -503,10 +506,9 @@ def __init__(self, n, p, *args, **kwargs):
         self.mean = self.n * self.p
         mode = tt.cast(tt.round(self.mean), 'int32')
         diff = self.n - tt.sum(mode, axis=-1, keepdims=True)
-        inc_bool_arr = tt.as_tensor_variable(lst) < diff
-        mode = tt.inc_subtensor(mode[inc_bool_arr.nonzero()], 1)
-        dec_bool_arr = tt.as_tensor_variable(lst) < -diff
```
Comment (dlovell) [inline pymc3/distributions/multivariate.py:508]: FYI: removing the decrement will cause this
```
import pymc3 as pm
N = 10
with pm.Model() as model:
    m = pm.Multinomial('m', N, [.26, .26, .26, .22], shape=(4,))
sum = m.distribution.mode.eval().sum()
assert sum == N, N-sum
```
to fail like this
```
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-10-9404ca61e1d5> in <module>()
----> 1 assert sum == N, N-sum

AssertionError: -1
```

Should I add this as a test?

Comment (junpenglao) [reply pymc3/distributions/multivariate.py:508]: Thanks, I think it should be fine now as I added `inc_bool_arr = tt.abs_(diff) > 0`

Comment (dlovell) [reply pymc3/distributions/multivariate.py:508]: oops, missed that

Comment (junpenglao) [reply pymc3/distributions/multivariate.py:508]: I edited a test for testing the case you mentioned as well.

**Comment by junpenglao** [comment]
This should fix #2507 and #2550. Ready when test pass.

File: pymc3/distributions/multivariate.py [hunk 2]
```diff
@@ -473,28 +473,31 @@ class Multinomial(Discrete):
...
-        be non-negative and sum to 1 along the last axis. They will be automatically
-        rescaled otherwise.
+        be non-negative and sum to 1 along the last axis. They will be 
+        automatically rescaled otherwise.
     """
 
     def __init__(self, n, p, *args, **kwargs):
         super(Multinomial, self).__init__(*args, **kwargs)
 
         p = p / tt.sum(p, axis=-1, keepdims=True)
```
Comment (ColCarroll) [inline pymc3/distributions/multivariate.py:487]: unrelated to the PR, but does `p` need to be scaled here if it is also scaled in the `._random` method?

File: pymc3/distributions/multivariate.py [hunk 3]
```diff
@@ -473,28 +473,31 @@ class Multinomial(Discrete):
...
             m = self.shape[-2]
             try:
                 assert n.shape == (m,)
-            except AttributeError:
-                n *= tt.ones(m)
+            except (AttributeError, AssertionError):
+                n = n * tt.ones(m)
+            self.n = tt.shape_padright(n)
+            self.p = p if p.ndim > 1 else tt.shape_padleft(p)
+        elif n.ndim == 1:
```
Comment (ColCarroll) [inline pymc3/distributions/multivariate.py:498]: it looks a little funny that the conditions here are `if len(self.shape) > 1: ... elif n.ndim == 1: ... else: ...`  is there a reason that both conditions can't be assertions about `self.shape`, or both about `n.ndim`?

**Review by ColCarroll — APPROVED** [review]

**Comment by ColCarroll** [comment]
Thanks -- this was tricky!

File: pymc3/distributions/multivariate.py (hunk 2 above)
Comment (junpenglao) [reply pymc3/distributions/multivariate.py:487]: I think so, just to make sure the logp is correct.
The rescaling in `._random` is more to dealt with the precision issue.

File: pymc3/distributions/multivariate.py (hunk 3 above)
Comment (junpenglao) [reply pymc3/distributions/multivariate.py:498]: Yes, `self.shape` doesnt work when there is `Observed`. I can't find a way to overwrite the default, so instead I am checking `n.dim` to interprete the shape.

I think we should refactor it to have a stricter shape (like what you said on twitter, `n` and `p` being both 2D and Observed being 2D or 3D). Something like:
```
n -> (k, 1)
p -> (m, k)
observerd -> (m, k) or (n, m, k)
```

Comment (ColCarroll) [reply pymc3/distributions/multivariate.py:498]: I was poking around, and there doesn't seem to be a theano `atleast_2d`.  The best solution would be for `numpy` or `scipy` to support such vectorized sampling...

**Pull request merged**